### ISSUE TITLE
Port S4 to functorch

### ISSUE THREAD

**Issue by zou3519** [pytorch/functorch#444]
Should be a good test case. https://github.com/srush/annotated-s4

**Comment by lucidrains** [comment]
@zou3519 @Chillee what's the status on functorch? are there any blockers for redoing s4 in pytorch? i may be up for taking this on

**Comment by Chillee** [comment]
@lucidrains None in principle...  Main things that might be tricky are implementing the equivalent of Flax's `nn.vmap` (which basically batches the parameter initialization as well as calls vmap on the function inputs). One thing we're currently not super certain about is the appropriate API for dealing with modules and vmap.

The other potential blocker is whether the JIT compilation (i.e. there's a KeOps style pattern) would be necessary for getting reasonable perf :) We have some approaches that might resolve this, but it's possible it'll take a little bit more effort.

TL;DR: Nothing massive in particular (except maybe compilation), probably mainly just worried that you'd run into some failure cases and think poorly of functorch :^)

**Comment by lucidrains** [comment]
ohh darn, you all should just hire the keops guys, i'm telling you lol

**Comment by lucidrains** [comment]
ok noted, thanks for letting me know
